src/field/RungeKuttaStepper.hh

```diff
@@ -0,0 +1,83 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file RungeKuttaStepper.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Range.hh"
+#include "base/Types.hh"
+#include "base/Macros.hh"
+//#include "FieldEquation.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * The 4th order classical Runge-Kutta stepper
+ *
+ * This method estimates the updated state from an initial state and slope
+ * estimate, with fourth-order accuracy.
+ *
+ * For a magnetic field equation \em f with state \em x, which includes
+ * position and momentum but may also include time and spin:
+ * \f[
+ *  XXX TODO: document steps here in latex math
+ * \f]
+ */
+template<class FieldEquation_T>
+class RungeKuttaStepper
+{
+    //@{
+    //! Type aliases
+    using OdeState = typename FieldEquation_T::OdeState;
+    //@}
+
+  public:
+    //! Construct with the equation of motion
+    inline CELER_FUNCTION RungeKuttaStepper(FieldEquation_T& equation);
+
+    //! Adaptive step size control
+    CELER_FUNCTION auto operator()(real_type       step,
+                                   const OdeState& beg_state,
+                                   const OdeState& beg_slope) -> OdeState;
```

Comment:
This class is doing a lot more than just acting like a function. If you've got a function-like interface, but also extra methods, it needs to be really clear what the expected calling order is  --  ideally so clear that you cannot compile an out-of-order result. If you expect that anything calling this class as a function needs to know `error` or `distance_chord` then they should return that as part  of the function call. If the chord doesn't need to be evaluated always, you can return a temporary class that encapsulates the result and the ability to lazily evaluate the chord.